File: 39. Combination Sum.swift
class Solution {
    func combinationSum(_ candidates: [Int], _ target: Int) -> [[Int]] {
        
        // 排序 + 消除重複
        let trimCandidates = removeDuplicates(candidates)
        
        // 定義：找出以 combination 為 root 的 所有results
        var results = [[Int]]()
        dfsToGetResults(trimCandidates, [], 0, target, &results)

        return results
    }
    
    // 定義：找出以 combination 為 root 的 所有results
    private func dfsToGetResults(_ candidates: [Int], _ root: [Int], _ startIndex: Int, _ target: Int, _ results: inout [[Int]]) {
        // 出口
        if root.reduce(0, +) == target {
            results.append(root)
            return
        }

        // 拆解
        for index in startIndex ..< candidates.count {
            
            if root.reduce(0, +) + candidates[index] > target {
                return
            }
            
            var nextRoot = root
            nextRoot.append(candidates[index])
            dfsToGetResults(candidates, nextRoot, index, target, &results)
        }
    }
    
    
    private func removeDuplicates(_ candidates: [Int]) -> [Int] {
        var result = [Int]()
        
        for item in candidates.sorted() {
            if result.last != item {
                result.append(item)
            }
        }
        return result
    }
}

File: 39. Combination Sum.py
import copy

class Solution:

    def removeDuplication(self, nums):
        nums.sort()

        results = []
        results.append(nums[0])

        for num in nums:
            if num != results[-1]:
                results.append(num)

        return results

    def sumOfCombination(self, combination):
        sum = 0
        for num in combination:
            sum += num

        return sum

    def dfs(self, nums, startIndex, target, tempCombination, results):
        if self.sumOfCombination(tempCombination) == target:
            temp = copy.deepcopy(tempCombination)
            results.append(temp)
            return

        if self.sumOfCombination(tempCombination) > target:
            return

        for i in range(startIndex, len(nums)):
            tempCombination.append(nums[i])
            self.dfs(nums, i, target, tempCombination, results)
            tempCombination.pop()

    def combinationSum(self, candidates, target):
        """
        :type candidates: List[int]
        :type target: int
        :rtype: List[List[int]]
        """

        if candidates is None or len(candidates) == 0:
            return

        # 去重
        nums = self.removeDuplication(candidates)

        tempCombination = []
        results = []

        self.dfs(nums, 0, target, tempCombination, results)

        return results
